In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf
from sklearn.datasets import fetch_covtype
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, BatchNormalization, Activation
import matplotlib.pyplot as plt

# Mandatory Seeding
np.random.seed(42)
tf.random.set_seed(42)

In [ ]:


# 2. Load the dataset
cov = fetch_covtype()
X = cov.data
y = cov.target

# Optional for quick testing during development:
# X = X[:10000]
# y = y[:10000]

print("--- Basic Dataset Information ---")
print("X shape:", X.shape)
print("y shape:", y.shape)
print("Number of classes:", len(np.unique(y)))
print("Class labels:", np.unique(y))

print("\n--- Class Distribution ---")
unique, counts = np.unique(y, return_counts=True)
for cls, cnt in zip(unique, counts):
    print(f"Class {cls}: {cnt}")

print("\n--- Feature Statistics (first few rows) ---")
df_stats = pd.DataFrame(X)
print(df_stats.describe().T.head(10))

--- Basic Dataset Information ---
X shape: (581012, 54)
y shape: (581012,)
Number of classes: 7
Class labels: [1 2 3 4 5 6 7]

--- Class Distribution ---
Class 1: 211840
Class 2: 283301
Class 3: 35754
Class 4: 2747
Class 5: 9493
Class 6: 17367
Class 7: 20510

--- Feature Statistics (first few rows) ---
      count         mean          std     min     25%     50%     75%     max
0  581012.0  2959.365301   279.984734  1859.0  2809.0  2996.0  3163.0  3858.0
1  581012.0   155.656807   111.913721     0.0    58.0   127.0   260.0   360.0
2  581012.0    14.103704     7.488242     0.0     9.0    13.0    18.0    66.0
3  581012.0   269.428217   212.549356     0.0   108.0   218.0   384.0  1397.0
4  581012.0    46.418855    58.295232  -173.0     7.0    30.0    69.0   601.0
5  581012.0  2350.146611  1559.254870     0.0  1106.0  1997.0  3328.0  7117.0
6  581012.0   212.146049    26.769889     0.0   198.0   218.0   231.0   254.0
7  581012.0   223.318716    19.768697     0.0   213.0   226.0   237.0   

In [ ]:
from tensorflow.keras.utils import to_categorical

# 3. Convert target labels from 1..7 to 0..6, then one-hot encode
y = y - 1
y_cat = to_categorical(y, num_classes=7)

# 4. Split into training and testing sets (80/20)
X_train, X_test, y_train, y_test = train_test_split(
    X, y_cat,
    test_size=0.2,
    random_state=seed_value,
    stratify=y
)

print("Train shape:", X_train.shape, y_train.shape)
print("Test shape:", X_test.shape, y_test.shape)

Train shape: (464809, 54) (464809, 7)
Test shape: (116203, 54) (116203, 7)


In [ ]:
# Standardize the features (recommended for ANN training)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
#  Model 1
model_1 = Sequential([
    Dense(10, activation='relu', input_dim=X_train.shape[1]),
    Dense(7, activation='softmax')
])

model_1.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

print("\nTraining Model 1...")
history_1 = model_1.fit(
    X_train, y_train,
    epochs=50,
    batch_size=256,
    verbose=1
)

print("\n--- Model 1 Results ---")
loss_1, acc_1 = model_1.evaluate(X_test, y_test, verbose=0)
print(f"Final Test Accuracy (Model 1): {acc_1:.4f}")


Training Model 1...
Epoch 1/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.6695 - loss: 0.8151
Epoch 2/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7336 - loss: 0.6437
Epoch 3/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.7396 - loss: 0.6195
Epoch 4/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7417 - loss: 0.6097
Epoch 5/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7437 - loss: 0.6047
Epoch 6/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.7452 - loss: 0.6018
Epoch 7/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.7459 - loss: 0.5996
Epoch 8/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7466 - loss: 0.5972
Epoch 9/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.7471 - loss: 0.5945
Epoch 10/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7477 - loss: 0.5924
Epoch 11/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7479 - loss: 0.5911
Epoch 12

In [ ]:
# Model 2
model_2 = Sequential([
    Dense(64, activation='relu', input_dim=X_train.shape[1]),
    Dense(32, activation='relu'),
    Dense(7, activation='softmax')
])

model_2.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

print("\nTraining Model 2...")
history_2 = model_2.fit(
    X_train, y_train,
    epochs=50,
    batch_size=256,
    verbose=1
)

print("\n--- Model 2 Results ---")
loss_2, acc_2 = model_2.evaluate(X_test, y_test, verbose=0)
print(f"Final Test Accuracy (Model 2): {acc_2:.4f}")


Training Model 2...
Epoch 1/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 17s 6ms/step - accuracy: 0.7417 - loss: 0.6172
Epoch 2/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 13s 2ms/step - accuracy: 0.7811 - loss: 0.5107
Epoch 3/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.7964 - loss: 0.4775
Epoch 4/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8066 - loss: 0.4561
Epoch 5/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8144 - loss: 0.4399
Epoch 6/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8207 - loss: 0.4265
Epoch 7/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.8260 - loss: 0.4155
Epoch 8/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8301 - loss: 0.4073
Epoch 9/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8335 - loss: 0.4003
Epoch 10/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8361 - loss: 0.3939
Epoch 11/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8389 - loss: 0.3879
Epoch 1

In [ ]:
# Manual calculation of trainable parameters for Model 2
input_features = 54
hidden1 = 64
hidden2 = 32
output_classes = 7

params_input_to_h1 = input_features * hidden1 + hidden1
params_h1_to_h2 = hidden1 * hidden2 + hidden2
params_h2_to_out = hidden2 * output_classes + output_classes

total_params = params_input_to_h1 + params_h1_to_h2 + params_h2_to_out

print("--- Manual Parameter Count for Model 2 ---")
print(f"Input -> Hidden1: {input_features}*{hidden1} + {hidden1} = {params_input_to_h1}")
print(f"Hidden1 -> Hidden2: {hidden1}*{hidden2} + {hidden2} = {params_h1_to_h2}")
print(f"Hidden2 -> Output: {hidden2}*{output_classes} + {output_classes} = {params_h2_to_out}")
print(f"Total Trainable Parameters: {total_params}")

--- Manual Parameter Count for Model 2 ---
Input -> Hidden1: 54*64 + 64 = 3520
Hidden1 -> Hidden2: 64*32 + 32 = 2080
Hidden2 -> Output: 32*7 + 7 = 231
Total Trainable Parameters: 5831


In [ ]:
#  Model 2 with Dropout and BatchNormalization
# BatchNormalization is typically placed after the Dense layer and before the activation.
# For that reason, Dense layers here use use_bias=False.

model_2_reg = Sequential([
    Dense(64, use_bias=False, input_dim=X_train.shape[1]),
    BatchNormalization(),
    Activation('relu'),
    Dropout(0.2),

    Dense(32, use_bias=False),
    BatchNormalization(),
    Activation('relu'),
    Dropout(0.2),

    Dense(7, activation='softmax')
])

model_2_reg.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

print("\nTraining Regularized Model 2...")
history_2_reg = model_2_reg.fit(
    X_train, y_train,
    epochs=50,
    batch_size=256,
    verbose=1
)

print("\n--- Regularized Model 2 Results ---")
loss_2_reg, acc_2_reg = model_2_reg.evaluate(X_test, y_test, verbose=0)
print(f"Final Test Accuracy (Model 2 + Dropout + BN): {acc_2_reg:.4f}")


Training Regularized Model 2...
Epoch 1/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.7013 - loss: 0.7244
Epoch 2/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.7422 - loss: 0.6044
Epoch 3/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.7542 - loss: 0.5737
Epoch 4/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.7609 - loss: 0.5578
Epoch 5/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.7646 - loss: 0.5467
Epoch 6/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.7682 - loss: 0.5371
Epoch 7/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.7725 - loss: 0.5304
Epoch 8/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.7740 - loss: 0.5262
Epoch 9/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.7758 - loss: 0.5217
Epoch 10/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.7769 - loss: 0.5190
Epoch 11/50
1816/1816 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.7799 - loss: 0.5

Model 2 generally achieves significantly higher accuracy than Model 1

Adding another hidden layer gives the network a hierarchical composition of features, so it can model interactions between elevation, distance, soil, and wilderness indicators more effectively than a single hidden layer. That extra composition is the practical benefit of depth in a feedforward ANN

Hidden layer 1: `(54 × 64) + 64 = 3,520`  
Hidden layer 2: `(64 × 32) + 32 = 2,080`  
Output layer: `(32 × 7) + 7 = 231`  

Total trainable parameters: `3,520 + 2,080 + 231 = 5,831`

Combine those simple features into complex, non-linear representations. Since forest cover types aren't separated by straight lines in a 54-dimensional space, the extra "curviness" provided by more ReLU activations helps the model fit the decision boundaries

Dropout randomly zeros activations during training, which reduces co-adaptation and usually improves generalization when the model starts to overfit. BatchNormalization keeps activations better scaled, which often makes optimization faster and more stable
